source/base/quadrature_lib.cc

```diff
@@ -774,10 +774,9 @@ QSorted<dim>::QSorted(const Quadrature<dim> &quad)
 
   std::sort(permutation.begin(),
             permutation.end(),
-            std::bind(&QSorted<dim>::compare_weights,
-                      std::ref(*this),
-                      std::placeholders::_1,
-                      std::placeholders::_2));
+            [this](const unsigned int x, const unsigned int y) {
```

Comment:
question: Why do you capture ``this`` like this explicitly? I haven't seen that used in deal.II before. I always just use ``[&]`` if something needs to be captured.

Reply:
Mainly to be explicit what needs to be captured. @bangerth liked that better in https://github.com/dealii/dealii/pull/8818#discussion_r326796217 and we also don't need to always capture by reference (we don't use references for small types, but copy instead).

Reply:
okay, I didn't see that conversation. Fine with me.

Reply:
To me the question is not capturing `this` but some of the other examples you have below where you list several variables to be captured: While not wrong, I feel one loses some of the charm of lambdas if one still has to declare which variables where used. As you said in https://github.com/dealii/dealii/pull/8818#discussion_r326817434 the compiler will typically get the scope of what is needed right. Thus my question: Do we want to encourage listing what gets captured rather than simply `[&]`. I'm certainly guilty of breaking that rule, so I guess it would be good to make a style decision at this point.

Reply:
Personally I like to use `[&]`. However, sometimes I prefer to explicitly specify what data structures to capture (which is in general a few) to make it clear to myself (and the reader) what the actual inputs are to the function. I think what to do should be the decision of the programmer (on the user hand we could definitely replace `[this]` by `[&]`).

Reply:
I could argue either way. `[&]` is short but not necessarily concise and listing everything might be precise, but overly verbose. I would be fine to encourage capture by reference by default, but also listing variables you want to copy explicitly. If the capture default is not used, I don't see any reason to use it, i.e., I prefer `[this]` over `[&]` if only `this` is captured.

For me, the capture list is not that much of a problem. They are typically much shorter than the declaration of the arguments.

Anyway, I am happy to follow any consensus we find.

Reply:
I like the flexible definition, so if someone wants to specify more than `[&]` then that is find as to transport an intent, and at the same time not require the verbose way in all circumstances.

Reply:
For the record, I don't feel strongly either way. We've generally erred on the side of being explicit about stuff in deal.II, but I don't feel like we *need* to be. 

In cases where we really just capture one or two variables, it might be nice to say so, but I wouldn't advocate for listing 5 or 8 variables in cases where we have a substantial amount of code in a lambda.

Do we actually need a policy, or could we leave the current set of patches as they are and just leave it to the authors in future cases to choose whichever way they prefer? We could of course try to be consistent, but we don't have *that* many lambdas that it's really going to matter. (Or maybe that counts as "famous last words"?)

Reply:
For reference, because someone mentioned copying small objects above: When you write
```
  [this]() {...}
```
or
```
  [=this]() {...}
```
then in both cases is the `this` *pointer* copied -- which is equivalent to capturing the `*this` object by *reference*.

Reply:
For the sake of completeness,
```
 [=, this]() {...}
```
is only supported from C++20 on. 
If implicitly captured, `(*this)` is captured by reference even if the capture default is `=`.
From C++17 on
```
 [*this]() {...}
```
can be used to copy the whole object.